changefeedccl: support end time option for changefeeds

sherman-grewal opened the pull request:
changefeedccl: Support end_time and initial_scan_only
options for changefeeds

Currently, users cannot indicate when they would like
their changefeeds to stop running. This becomes an
issue when a user would like to run a changefeed
until a certain time, or would like to run an initial
scan only.

In this PR we introduce the capability that will allow
users to indicate a time at which they would like their
changefeed to stop running. This can be achieved by
providing a timestamp for the 'end_time' option.

Moreover, users can also indicate if they want to perform
an initial scan only. This can be achieved by providing the
'initial_scan_only' option.

When the end time is reached, or the initial scan is
complete (when given the initial_scan_only option), we end
the changefeed job with a successful status.

Release note (enterprise change): Allow users to provide
an end time for changefeeds through the 'end_time' option.
When this option is provided, the changefeed will run until
it has reached the end time, and then the changefeed job will
end with a successful status code. Furthermore, we now provide
an 'initial_scan_only' option. When this option is set, the
changefeed job will end with a successful status code
immediately after the initial scan completed.

Release Justification: This feature is low danger, yet
highly desired from users.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/78199)
<!-- Reviewable:end -->

miretskiy commented:
haven't looked closely, but this looks pretty much what I expected

samiskin commented on pkg/ccl/changefeedccl/changefeed_test.go:
```diff
@@ -5612,6 +5632,68 @@ func TestChangefeedCaseInsensitiveOpts(t *testing.T) {
+		sqlDB.Exec(t, "CREATE TABLE foo (a INT PRIMARY KEY)")
+		sqlDB.Exec(t, "INSERT INTO foo VALUES (1), (2), (3)")
+
+		feed := feed(t, f, "CREATE CHANGEFEED FOR foo WITH initial_scan_only")
```
In both of these tests it'd be nice to have an extra `INSERT` statement right after this `CREATE CHANGEFEED` / end time and then ensure that no kv events other than the initial scan events were emitted.  

`assertPayloads` alone wouldn't be able to assert "nothing extra was sent" though, one option would be a separate goroutine that consumes from the sink ([ex](https://github.com/cockroachdb/cockroach/blob/1eafcc40c352ae9a4ffb9bddf3999a9cf002a4ad/pkg/ccl/changefeedccl/changefeed_test.go#L1219)) and increments a counter for the number of events its seen and ensuring that that number didn't exceed 3.

sherman-grewal replied:
Thanks for this suggestion! I was actually looking to carry out a test similar to the one you described above, but I wasn't sure if it was possible. I will look into improving these tests though 💯

miretskiy reviewed (approved):
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @samiskin, and @sherman-grewal)

___
*[-- commits, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MymXg5-5FBfvzkvATZs):*
you need to have release justification as well.  Also, pls copy this description to gh pr description
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 5646 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MymfHn66XplRJmu9-o2:-MymfHn66XplRJmu9-o3:b-a8vs9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/1eafcc40c352ae9a4ffb9bddf3999a9cf002a4ad/pkg/ccl/changefeedccl/changefeed_test.go#L5646)):*
> ```Go
> 
> 		var tsAfterInitialInsert string
> 		sqlDB.QueryRow(t, "SELECT cluster_logical_timestamp()").Scan(&tsAfterInitialInsert)
> ```

please insert few more rows here (still expecting to see 1,2,3 only)
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 5648 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MymfWkJ-Y56BLR35gu1:-MymfWkJ-Y56BLR35gu2:b-a3fdx0) ([raw file](https://github.com/cockroachdb/cockroach/blob/1eafcc40c352ae9a4ffb9bddf3999a9cf002a4ad/pkg/ccl/changefeedccl/changefeed_test.go#L5648)):*
> ```Go
> 		sqlDB.QueryRow(t, "SELECT cluster_logical_timestamp()").Scan(&tsAfterInitialInsert)
> 
> 		feed := feed(t, f, "CREATE CHANGEFEED FOR foo WITH end_time = $1", tsAfterInitialInsert)
> ```

would be nice to add a test that has both the cursor and an end time.


<!-- Sent from Reviewable.io -->

samiskin commented on pkg/ccl/changefeedccl/kvfeed/kv_feed.go:
```diff
@@ -215,8 +245,21 @@ func (f *kvFeed) run(ctx context.Context) (err error) {
 			return err
 		}
 
+		if f.initialScanOnly {
```
Would it work to just treat `initialScanOnly` as `EndTime == initialHighWater` (when we have `withInitialBackfill`) and not need the extra parameter at all?

sherman-grewal replied:
When we indicate `initialScanOnly`, we do not want the rangefeed to start after the initial scan has run. However, if we were to go through with the idea you proposed, the rangefeed would turn on as we would reach `runUntilTableEvent`. This is why I think we need the two parameters.

samiskin replied:
I don't necessarily mean removing the if statement against running `runUntilTableEvent`, just mean not needing the extra `InitialScanOnly` property that gets passed around and instead checking if your highwater is already at the endtime

sherman-grewal replied:
ahh ok, yeah I can go through with that instead, thanks for the suggestion!

samiskin reviewed (commented):
Great work!

miretskiy reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @samiskin, and @sherman-grewal)

___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 583 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MynOn5j5TrBYmDiwkP8:-MynOn5k1sxylvz9FIhw:bnsrmnr) ([raw file](https://github.com/cockroachdb/cockroach/blob/1eafcc40c352ae9a4ffb9bddf3999a9cf002a4ad/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L583)):*
> ```Go
> 				// Break out of the ctxgroup by returning a sentinel error.
> 				return &errEndTimeReached{endTime: endTime}
> 			}
> ```

I'm not sure this is correct actually; we must make sure we reach "scan boundary
" before we error out.


<!-- Sent from Reviewable.io -->

sherman-grewal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @samiskin, and @sherman-grewal)

___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 583 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MynOn5j5TrBYmDiwkP8:-Myrotti2n0HJQ6Yoh0J:b-vajga2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1eafcc40c352ae9a4ffb9bddf3999a9cf002a4ad/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L583)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I'm not sure this is correct actually; we must make sure we reach "scan boundary
" before we error out.
</blockquote></details>

Just want to clarify the comment above and the changes I made to address this comment:

For those who are wondering, previously I would check if the endTime was less than or equal to the frontier at the end of the `copyEvent` function. Although this would ensure that the changefeed would succeed once we reach the endTime, the issue with this approach is that we would not skip resolved events once a span has reached the endTime boundary. If you look closely at `applyScanBoundary` we skip resolved events every time a span has reached the boundary, and we do this as an optimization for performance. In order to carry out this optimization for performance when checking the endTime boundary, we inject the endTimeReached error into the scanBoundary error whenever there are no table events. This way, the endTimeReached error goes through the same flow of checkForScanBoundary, applyScanBoundary, addEntry.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
> Reviewed 1 of 8 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @samiskin, and @sherman-grewal)

___
*[-- commits, line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MyrpL7N7Q-xfOKSrSLP):*
Can you adjust PR description to indicate that alter stmt for these options is not supported
at this time.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 364 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MyrpbxaDEXYNXifckk5:-MyrpbxaDEXYNXifckk6:b7fdjmq) ([raw file](https://github.com/cockroachdb/cockroach/blob/02754f757ee5d7c491e40d03618e1d9c6b90d28d/pkg/ccl/changefeedccl/changefeed_processors.go#L364)):*
> ```Go
> 
> 	if schemaChangePolicy == changefeedbase.OptSchemaChangePolicyIgnore || initialScanOnly {
> 		sf = schemafeed.DoNothingSchemaFeed
> ```

ohh, that's nice.
___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 583 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MynOn5j5TrBYmDiwkP8:-MysBzcB4svTkGjv_UbA:b8v7wjk) ([raw file](https://github.com/cockroachdb/cockroach/blob/1eafcc40c352ae9a4ffb9bddf3999a9cf002a4ad/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L583)):*
<details><summary><i>Previously, sherman-grewal (Sherman) wrote…</i></summary><blockquote>

Just want to clarify the comment and the changes I made to address the above comment:

For those who are wondering, previously I would check if the endTime was less than or equal to the frontier at the end of the `copyEvent` function. Although this would ensure that the changefeed would succeed once we reach the endTime, the issue with this approach is that we would not skip resolved events once a span has reached the endTime boundary. If you look closely at `applyScanBoundary` we skip resolved events every time a span has reached the boundary, and we do this as an optimization for performance. In order to carry out this optimization for performance when checking the endTime boundary, we inject the endTimeReached error into the scanBoundary error whenever there are no table events. This way, the endTimeReached error goes through the same flow of checkForScanBoundary, applyScanBoundary, addEntry.
</blockquote></details>

I think the issue with the previous approach was that we return "time reached" error the first first time the fastest range event reaches that time.  But what about other ranges that are behind? We have to make sure they all reach the frontier.  While slower spans catchup, we must make sure that we skip events for the ranges that have already reached the frontier.
___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 443 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-Mys8j-v5DwqBUVHZgmo:-Mys8j-v5DwqBUVHZgmp:bmbt2de) ([raw file](https://github.com/cockroachdb/cockroach/blob/02754f757ee5d7c491e40d03618e1d9c6b90d28d/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L443)):*
> ```Go
> 
> type errBoundaryReached interface {
> 	Error() string
> ```

perhaps embed `error` interface (instead of Error() method).
___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 523 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MysBlsN0UgyScEj8Sh-:-MysBlsN0UgyScEj8Sh0:b6winc5) ([raw file](https://github.com/cockroachdb/cockroach/blob/02754f757ee5d7c491e40d03618e1d9c6b90d28d/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L523)):*
> ```Go
> 			if len(nextEvents) > 0 {
> 				scanBoundary = &errTableEventReached{nextEvents[0]}
> 			} else if !endTime.IsEmpty() {
> ```

should we check for `!endTime.IsEmpty() && scanBoundary == nil`?


<!-- Sent from Reviewable.io -->

Commit:
changefeedccl: Support end_time and initial_scan_only
options for changefeeds

Currently, users cannot indicate when they would like
their changefeeds to stop running. This becomes an
issue when a user would like to run a changefeed
until a certain time, or would like to run an initial
scan only.

In this PR we introduce the capability that will allow
users to indicate a time at which they would like their
changefeed to stop running. This can be achieved by
providing a timestamp for the 'end_time' option.

Moreover, users can also indicate if they want to perform
an initial scan only. This can be achieved by providing the
'initial_scan_only' option.

When the end time is reached, or the initial scan is
complete (when given the initial_scan_only option), we end
the changefeed job with a successful status.

Note that in this PR we disallow the alteration of the
end_time and initial_scan_only options. This is so that
it is easier to backport this PR. We will later open up
another PR that supports the alteration of these options.

Release note (enterprise change): Allow users to provide
an end time for changefeeds through the 'end_time' option.
When this option is provided, the changefeed will run until
it has reached the end time, and then the changefeed job will
end with a successful status code. Furthermore, we now provide
an 'initial_scan_only' option. When this option is set, the
changefeed job will end with a successful status code
immediately after the initial scan completed.

Release Justification: This feature is low danger, yet
highly desired from users.

sherman-grewal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/78199)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, and @samiskin)

___
*[-- commits, line 31 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MymXg5-5FBfvzkvATZs):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

you need to have release justification as well.  Also, pls copy this description to gh pr description
</blockquote></details>

Done.
___
*[-- commits, line 31 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MyrpL7N7Q-xfOKSrSLP):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Can you adjust PR description to indicate that alter stmt for these options is not supported
at this time.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_processors.go, line 364 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MyrpbxaDEXYNXifckk5:-MysJ_KlBzV3BYgdjmX-:b32vhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/02754f757ee5d7c491e40d03618e1d9c6b90d28d/pkg/ccl/changefeedccl/changefeed_processors.go#L364)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

ohh, that's nice.
</blockquote></details>

:)
___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 583 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MynOn5j5TrBYmDiwkP8:-MysJehYCwOD2VUNG8l1:b-yflw11) ([raw file](https://github.com/cockroachdb/cockroach/blob/1eafcc40c352ae9a4ffb9bddf3999a9cf002a4ad/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L583)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think the issue with the previous approach was that we return "time reached" error the first first time the fastest range event reaches that time.  But what about other ranges that are behind? We have to make sure they all reach the frontier.  While slower spans catchup, we must make sure that we skip events for the ranges that have already reached the frontier.
</blockquote></details>

Yeah that's a good point, thanks for clarifying the issue.
___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 443 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-Mys8j-v5DwqBUVHZgmo:-MysJbGI8ZMMI1ePgonq:bt4ox15) ([raw file](https://github.com/cockroachdb/cockroach/blob/02754f757ee5d7c491e40d03618e1d9c6b90d28d/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L443)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

perhaps embed `error` interface (instead of Error() method).
</blockquote></details>

Done. Thanks for the suggestion!
___
*[pkg/ccl/changefeedccl/kvfeed/kv_feed.go, line 523 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MysBlsN0UgyScEj8Sh-:-MysJck55BOPN7BdJJ77:b-kd09o8) ([raw file](https://github.com/cockroachdb/cockroach/blob/02754f757ee5d7c491e40d03618e1d9c6b90d28d/pkg/ccl/changefeedccl/kvfeed/kv_feed.go#L523)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

should we check for `!endTime.IsEmpty() && scanBoundary == nil`?
</blockquote></details>

Done. Thanks for the suggestion!!
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 5646 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MymfHn66XplRJmu9-o2:-MysJmmZ6HKXZCNQOQA_:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1eafcc40c352ae9a4ffb9bddf3999a9cf002a4ad/pkg/ccl/changefeedccl/changefeed_test.go#L5646)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

please insert few more rows here (still expecting to see 1,2,3 only)
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeed_test.go, line 5648 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/78199#-MymfWkJ-Y56BLR35gu1:-MysJn7A3GM8qLjKupwk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1eafcc40c352ae9a4ffb9bddf3999a9cf002a4ad/pkg/ccl/changefeedccl/changefeed_test.go#L5648)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

would be nice to add a test that has both the cursor and an end time.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

samiskin reviewed (approved):
💯 ✨

sherman-grewal commented:
bors r+
